# HaploQA_qtl2_comp
Input data received from HaploQA - https://haploqa.jax.org/, software platform that provides interpretation of Mouse Universal Genotyping array platforms (MUGA) genotype calls available from Neogen Genomics (Morgan et al., 2015).

Computation implementations based on R/qtl2 - Broman KW, Gatti DM, Simecek P, Furlotte NA, Prins P, Sen Åš, Yandell BS, Churchill GA (2018) R/qtl2: software for mapping quantitative trait loci with high-dimensional data and multi-parent populations. Genetics 211:495-502

The purpose of this project is to determine the accuracy of outputs from HaploQA and qtl2 generalized AIL (Advanced intercross lines) methods compared to that of the optimal qtl2 models (referred in pipeline as truth models), and decide which method is the best for performing haplotype reconstruction tasks

# Overview
input_data_prep_functions.R - Main script containing all pipelines and utility functions
HaploQA_qtl2_output.R - Script used to test the functions
app.R - Shiny app functions for genotype visualizations

draft_report.Rmd - Markdown report of the project (work in progress)

annotations_config.csv, founder_strains_table.csv - dictionaries used for mapping

# Usage
Overview of the programmatic outlines:
1. Data retrieval (selected samples from https://haploqa.jax.org/) and wrangling
2. Three major pipelines:
  a. 'haplotype_reconstruction_pipeline' - haplotype reconstructions for an entire cross (all individuals within the cross)
  b. 'sample_haplotype_reconstruction' - haplotype reconstructions for one individual
  c. 'truth_model_reconstruction' - haplotype reconstructions for optimal/truth models
3. metrics calculations - compare outputs from 'a' or 'b' above to that of 'c'


